File: Kartik/Program3/RailFenceCipher.c
#include <stdio.h>
#include <string.h>

// Function to encrypt plaintext using Rail Fence Cipher
void railFenceEncrypt(char plaintext[], int key) {
    int len = strlen(plaintext);
    char railMatrix[key][len];
    
    // Initialize rail matrix
    for (int i = 0; i < key; i++)
        for (int j = 0; j < len; j++)
            railMatrix[i][j] = '\0';
    
    // Fill the rail matrix
    int row = 0;
    int dir = 0; // 0 for down, 1 for up
    for (int i = 0; i < len; i++) {
        railMatrix[row][i] = plaintext[i];
        if (row == 0)
            dir = 0;
        else if (row == key - 1)
            dir = 1;
        if (dir == 0)
            row++;
        else
            row--;
    }
    
    // Read the rail matrix to get encrypted text
    printf("Encrypted text: ");
    for (int i = 0; i < key; i++)
        for (int j = 0; j < len; j++)
            if (railMatrix[i][j] != '\0')
                printf("%c", railMatrix[i][j]);
    printf("\n");

    // Now decrypt the encrypted text
    printf("Decrypted text: ");
    row = 0;
    dir = 0;
    for (int i = 0; i < len; i++) {
        printf("%c", railMatrix[row][i]);
        if (row == 0)
            dir = 0;
        else if (row == key - 1)
            dir = 1;
        if (dir == 0)
            row++;
        else
            row--;
    }
    printf("\n");
}

int main() {
    char text[100];
    int key;
    
    printf("Enter the plaintext: ");
    fgets(text, sizeof(text), stdin);
    printf("Enter the key (number of rails): ");
    scanf("%d", &key);
    
    
    // Remove newline character from text
    text[strcspn(text, "\n")] = '\0';
    
    railFenceEncrypt(text, key);
    
    return 0;
}

File: Kartik/Program2/RailFenceCipher.java
import java.util.*;

public class RailFenceCipher {
    public static String encrypt(String plaintext, int rails) {
        StringBuilder ciphertext = new StringBuilder();

        // Create the rail fence pattern
        String[] fence = new String[rails];
        for (int i = 0; i < rails; i++) {
            fence[i] = "";
        }

        int rail = 0;
        boolean down = true;
        for (char c : plaintext.toCharArray()) {
            fence[rail] += c;
            if (rail == 0) {
                down = true;
            } else if (rail == rails - 1) {
                down = false;
            }
            rail += down ? 1 : -1;
        }

        // Concatenate the fence rows to form the ciphertext
        for (String railString : fence) {
            ciphertext.append(railString);
        }

        return ciphertext.toString();
    }

    public static String decrypt(String ciphertext, int rails) {
        StringBuilder plaintext = new StringBuilder();

        // Create the rail fence pattern
        String[] fence = new String[rails];
        for (int i = 0; i < rails; i++) {
            fence[i] = "";
        }

        int[] railLengths = new int[rails];
        int rail = 0;
        boolean down = true;
        for (int i = 0; i < ciphertext.length(); i++) {
            railLengths[rail]++;
            if (rail == 0) {
                down = true;
            } else if (rail == rails - 1) {
                down = false;
            }
            rail += down ? 1 : -1;
        }

        // Fill the rail fence with the ciphertext
        int index = 0;
        for (int i = 0; i < rails; i++) {
            for (int j = 0; j < railLengths[i]; j++) {
                fence[i] += ciphertext.charAt(index++);
            }
        }

        // Read the plaintext from the rail fence
        rail = 0;
        down = true;
        for (int i = 0; i < ciphertext.length(); i++) {
            plaintext.append(fence[rail].charAt(0));
            fence[rail] = fence[rail].substring(1);
            if (rail == 0) {
                down = true;
            } else if (rail == rails - 1) {
                down = false;
            }
            rail += down ? 1 : -1;
        }

        return plaintext.toString();
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the PlainText to encrypt:_ ");
        String plaintext = sc.nextLine();
        System.out.println("Enter the Rail value:");
        int rails = sc.nextInt();

        // Encrypt the plaintext
        String encryptedText = encrypt(plaintext, rails);
        System.out.println("Encrypted text: " + encryptedText);

        // Decrypt the ciphertext
        String decryptedText = decrypt(encryptedText, rails);
        System.out.println("Decrypted text: " + decryptedText);
    }
}

